src/python/pants/backend/python/subsystems/setup.py

```diff
@@ -363,6 +311,82 @@ class PythonSetup(Subsystem):
         ),
         advanced=True,
     )
+    requirement_constraints = FileOption(
+        "--requirement-constraints",
+        default=None,
+        help=softwrap(
+            """
+            When resolving third-party requirements for your own code (vs. tools you run),
+            use this constraints file to determine which versions to use.
+
+            Mutually exclusive with `[python].enable_resolves`, which we generally recommend as an
+            improvement over constraints file, including due to less supply chain risk thanks to
+            `--hash` support.
+
+            See https://pip.pypa.io/en/stable/user_guide/#constraints-files for more
+            information on the format of constraint files and how constraints are applied in
+            Pex and pip.
+
+            This only applies when resolving user requirements, rather than tools you run
+            like Black and Pytest. To constrain tools, set `[tool].lockfile`, e.g.
+            `[black].lockfile`.
+            """
+        ),
+        advanced=True,
+        mutually_exclusive_group="lockfile",
+    )
+    resolve_all_constraints = BoolOption(
+        "--resolve-all-constraints",
+        default=True,
+        help=softwrap(
+            """
+            (Only relevant when using `[python].requirement_constraints.`) If enabled, when
+            resolving requirements, Pants will first resolve your entire
+            constraints file as a single global resolve. Then, if the code uses a subset of
+            your constraints file, Pants will extract the relevant requirements from that
+            global resolve so that only what's actually needed gets used. If disabled, Pants
+            will not use a global resolve and will resolve each subset of your requirements
+            independently.
+
+            Usually this option should be enabled because it can result in far fewer resolves.
+            """
+        ),
+        advanced=True,
+    )
+    no_binary = StrListOption(
+        "--no-binary",
+        help=softwrap(
+            """
+            Do not use binary packages (i.e., wheels) for these 3rdparty projects.
```

Comment:
The text is the same, but I added paragraph breaks. I also re-worded the last paragraph to be more accurate and concise